staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -1115,7 +1115,7 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 				} else {
 					for i, cr := range compResults {
 						expressionCost := getExpressionCost(cr, celContext)
-						if expressionCost > StaticEstimatedCostLimit {
+						if !opts.preexistingExpressions.rules.Has(schema.XValidations[i].Rule) && expressionCost > StaticEstimatedCostLimit {
```

Comment:
There are ~three things that inform the max cost of an expression:
1. the complexity of the expression itself
2. the characteristics (size, etc) of the data the expression references
3. the number of times it can be called

This looks to be exempting max cost controls based on 1 only (if the old object contained expression "X", we no longer error on the individual cost of the same expression in the new object, no matter the cost)

I *think* this means the following escapes cost limits:
1. User creates CRD with bounded items (list with `maxItems`, schema with fields with `maxLength`, etc) and some expression
2. User updates CRD keeping the same expression but expanding or dropping the bounds (dropping maxItems, etc)

Am I reading that right?

Reply:
Yes you are absolutely correct. Sorry for our rushed patch. Let me review the cost estimation algorithm and add tracking for the remainig.

Reply:
That makes sense. We would like to prevent users from updating other part of the CRD but leave expressions the same to skip the check. SO the skip needs to based on CRD version and under the condition of schema(under this version) remains the same. cc @jiahuif Does this make sense?

Reply:
Sounds good. Per offline discussion I can go ahead with a solution that compares each version, and suppress per-expression cost checks for unchanged versions.